lightning/src/ln/interactivetxs.rs

```diff
@@ -97,54 +97,190 @@ pub enum AbortReason {
 	InvalidTx,
 }
 
-#[derive(Debug)]
-pub struct TxInputWithPrevOutput {
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) struct InteractiveTxInput {
+	serial_id: SerialId,
 	input: TxIn,
 	prev_output: TxOut,
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) struct InteractiveTxOutput {
+	serial_id: SerialId,
+	tx_out: TxOut,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) struct ConstructedTransaction {
+	holder_is_initiator: bool,
+
+	inputs: Vec<InteractiveTxInput>,
+	outputs: Vec<InteractiveTxOutput>,
+
+	local_inputs_value_satoshis: u64,
+	local_outputs_value_satoshis: u64,
+
+	remote_inputs_value_satoshis: u64,
+	remote_outputs_value_satoshis: u64,
+
+	lock_time: AbsoluteLockTime,
+}
+
+impl ConstructedTransaction {
+	pub fn new(
+		holder_is_initiator: bool, inputs: Vec<InteractiveTxInput>,
+		outputs: Vec<InteractiveTxOutput>, lock_time: AbsoluteLockTime,
+	) -> Self {
+		let mut local_inputs_value_satoshis = 0u64;
+		let mut remote_inputs_value_satoshis = 0u64;
+		for input in &inputs {
+			if is_serial_id_valid_for_counterparty(holder_is_initiator, &input.serial_id) {
+				remote_inputs_value_satoshis =
+					remote_inputs_value_satoshis.saturating_add(input.prev_output.value);
+			} else {
+				local_inputs_value_satoshis =
+					local_inputs_value_satoshis.saturating_add(input.prev_output.value);
+			}
+		}
+
+		let mut local_outputs_value_satoshis = 0u64;
+		let mut remote_outputs_value_satoshis = 0u64;
+		for output in &outputs {
+			if is_serial_id_valid_for_counterparty(holder_is_initiator, &output.serial_id) {
+				remote_outputs_value_satoshis =
+					remote_outputs_value_satoshis.saturating_add(output.tx_out.value);
+			} else {
+				local_outputs_value_satoshis =
+					local_outputs_value_satoshis.saturating_add(output.tx_out.value);
+			}
+		}
+
+		Self {
+			holder_is_initiator,
+
+			inputs,
+			outputs,
+
+			local_inputs_value_satoshis,
+			local_outputs_value_satoshis,
+
+			remote_inputs_value_satoshis,
+			remote_outputs_value_satoshis,
+
+			lock_time,
+		}
+	}
+
+	pub fn weight(&self) -> u64 {
+		let inputs_weight =
+			self.inputs.iter().fold(0u64, |weight, InteractiveTxInput { prev_output, .. }| {
+				weight.saturating_add(estimate_input_weight(prev_output))
+			});
+		let outputs_weight =
+			self.outputs.iter().fold(0u64, |weight, InteractiveTxOutput { tx_out, .. }| {
+				weight.saturating_add(get_output_weight(&tx_out.script_pubkey))
+			});
+
+		let tx =
+			Transaction { version: 2, lock_time: self.lock_time, input: vec![], output: vec![] };
+
+		tx.weight().to_wu().saturating_add(inputs_weight).saturating_add(outputs_weight)
```

Comment:
Yeah no problem doing that! This is also really just an estimate (we don't have witness weights) so we are going to give wiggle room on the feerate after exchanging signatures anyway.

Reply:
I'll return `Weight` here but internally we'll still calculate with u64 since `Weight` only has `checked_add` and not `saturating_add` which would get messy.

Reply:
FWIW, `checked_add(...).unwrap(Weight::MAX)` should be the equivalent of `saturating_add`.

Reply:
FYI, if you need to string a bunch of these together, using `and_then` repeatedly with a final `unwrap_or` would make it a bit less messy.

Reply:
Thanks, I'm going to fix that up here. I keep forgetting `and_then` and friends.